### PR TITLE
service/iam: iam_principal_policy_simulation data source

### PR DESCRIPTION
**Pull request by apparentlymart**
This data source wraps the IAM policy simulation API. This was previously an API action with little utility in Terraform, but with the introduction of preconditions and postconditions in Terraform v1.2.3 it can be potentially useful as a way for a configuration to either pre-verify that it seems to be running with credentials that confer sufficient access or to self-check a policy it declares itself to get earlier warning if the policy seems insufficient for its intended purpose.

Unfortunately the IAM policy simulator is pretty low-level and requires the caller to figure out all of the relevant details of how a real AWS service would make requests to IAM at runtime in order to construct a fully-realistic simulation, but thankfully in practice it seems like authors could make do with relatively-simple "naive" simulations unless they know they are using more complex IAM policy features, such as custom conditions or interpolations. I included some hopefully-realistic examples in the documentation.

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Output from acceptance testing:

```
=== RUN   TestAccIAMPrincipalPolicySimulationDataSource
=== PAUSE TestAccIAMPrincipalPolicySimulationDataSource
=== CONT  TestAccIAMPrincipalPolicySimulationDataSource
--- PASS: TestAccIAMPrincipalPolicySimulationDataSource (23.77s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/iam        26.721s
```

### CODE REVIEW

File: internal/service/iam/principal_policy_simulation_data_source.go [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+			// Arguments
+			"action_names": {
+				Type:     schema.TypeSet,
+				Required: true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+				Description: `One or more names of actions, like "iam:CreateUser", that should be included in the simulation.`,
+			},
+			"caller_arn": {
```
Comment (DrFaust92) [inline internal/service/iam/principal_policy_simulation_data_source.go:26]: add validation for arn?

File: internal/service/iam/principal_policy_simulation_data_source.go [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+							Elem: &schema.Schema{
+								Type: schema.TypeString,
+							},
+							Description: `One or more values to assign to the context key, given as a string in a syntax appropriate for the selected value type.`,
+						},
+					},
+				},
+				Description: `Each block specifies one item of additional context entry to include in the simulated requests. These are the additional properties used in the 'Condition' element of an IAM policy, and in dynamic value interpolations.`,
+			},
+			"permissions_boundary_policies_json": {
```
Comment (DrFaust92) [inline internal/service/iam/principal_policy_simulation_data_source.go:58]: add validation for valid json?

File: internal/service/iam/principal_policy_simulation_data_source.go [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+				Type:     schema.TypeSet,
+				Optional: true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+				Description: `Additional permission boundary policies to use in the simulation.`,
+			},
+			"additional_policies_json": {
```
Comment (DrFaust92) [inline internal/service/iam/principal_policy_simulation_data_source.go:66]: add validation for valid json

File: internal/service/iam/principal_policy_simulation_data_source.go [hunk 4]
```diff
@@ -0,0 +1,326 @@
...
+				Type:     schema.TypeSet,
+				Optional: true,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+				},
+				Description: `Additional principal-based policies to use in the simulation.`,
+			},
+			"policy_source_arn": {
```
Comment (DrFaust92) [inline internal/service/iam/principal_policy_simulation_data_source.go:74]: add validation for valid arn

File: internal/service/iam/principal_policy_simulation_data_source.go [hunk 5]
```diff
@@ -0,0 +1,326 @@
...
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: `ARN of the principal (e.g. user, role) whose existing configured access policies will be used as the basis for the simulation. If you specify a role ARN here, you can also set caller_arn to simulate a particular user acting with the given role.`,
+			},
+			"resource_arns": {
```
Comment (DrFaust92) [inline internal/service/iam/principal_policy_simulation_data_source.go:79]: add validations for arn

File: internal/service/iam/principal_policy_simulation_data_source.go [hunk 6]
```diff
@@ -0,0 +1,326 @@
...
+					Type: schema.TypeString,
+				},
+				Description: `ARNs of specific resources to use as the targets of the specified actions during simulation. If not specified, the simulator assumes "*" which represents general access across all resources.`,
+			},
+			"resource_handling_option": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: `Specifies the type of simulation to run. Some API operations need a particular resource handling option in order to produce a correct reesult.`,
+			},
+			"resource_owner_account_id": {
```
Comment (DrFaust92) [inline internal/service/iam/principal_policy_simulation_data_source.go:92]: add validation for valid account id

File: internal/service/iam/principal_policy_simulation_data_source.go [hunk 7]
```diff
@@ -0,0 +1,326 @@
...
+				Type:        schema.TypeString,
+				Optional:    true,
+				Description: `An AWS account ID to use as the simulated owner for any resource whose ARN does not include a specific owner account ID. Defaults to the account given as part of caller_arn.`,
+			},
+			"resource_policy_json": {
```
Comment (DrFaust92) [inline internal/service/iam/principal_policy_simulation_data_source.go:97]: json validation

**Review by DrFaust92 — CHANGES_REQUESTED** [review]

**Review by ewbankkit — APPROVED** [review]
LGTM :rocket:.

```console
% make testacc TESTARGS='-run=TestAccIAMPrincipalPolicySimulationDataSource' PKG=iam
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/iam/... -v -count 1 -parallel 20  -run=TestAccIAMPrincipalPolicySimulationDataSource -timeout 180m
=== RUN   TestAccIAMPrincipalPolicySimulationDataSource_basic
=== PAUSE TestAccIAMPrincipalPolicySimulationDataSource_basic
=== CONT  TestAccIAMPrincipalPolicySimulationDataSource_basic
--- PASS: TestAccIAMPrincipalPolicySimulationDataSource_basic (27.03s)
PASS
ok  	github.com/hashicorp/terraform-provider-aws/internal/service/iam	32.824s
```

**Comment by ewbankkit** [comment]
@apparentlymart Thanks for the contribution :tada: :clap:.

**Pull request merged**